### ISSUE TITLE
Model has a GPU argument at load and predict

### ISSUE THREAD

**Issue by schmmd** [allenai/allennlp#1107]
Presently the [model class](https://github.com/allenai/allennlp/blob/master/allennlp/models/model.py) takes a `cuda_device` in both the [load method](https://github.com/allenai/allennlp/blob/cbe58971a15aaf6a80d3cc53b1e7c4a9ba4e089c/allennlp/models/model.py#L199) but so does the [forward method](https://github.com/allenai/allennlp/blob/cbe58971a15aaf6a80d3cc53b1e7c4a9ba4e089c/allennlp/models/model.py#L122).  Ideally calls to forward would, at least by default, use the same `cuda_device` that the model was loaded on.

Would it even work to load a model on one device but to make predictions on another?

**Comment by matt-gardner** [comment]
The `cuda_device` parameter in the `load` method applies to the model parameters, while the `cuda_device` parameter in `forward_on_instances` applies to the _data_.  Yes, these need to match.  You could conceivably have a multi-gpu model, though, and you need to know which device to put the data on.  It's possible there's a way to get around needing `cuda_device` in `forward_on_instances`, but if there is I'm not sure what it is.  In any event, we should improve the documentation to make it more obvious what's going on here.

**Comment by schmmd** [comment]
Right--my point is that since they need to match shouldn't the `cuda_device` on `forward_on_instances` just default to what was sent to `load`?  This would simplify usage.

**Comment by matt-gardner** [comment]
Great, just making sure I understood your issue.  I agree.

**Comment by schmmd** [comment]
Fixed in https://github.com/allenai/allennlp/pull/1151/files.

**Comment by larry0123du** [comment]
@matt-gardner could you elaborate a bit on how to use multiple cuda devices? Couldn't find any information out there.

**Comment by matt-gardner** [comment]
I don't actually know what the state of our multi-gpu support is, except I don't think it works in most cases. @DeNeutoy or @matt-peters know more than me here.

**Comment by matt-peters** [comment]
MultiGPU training generally works, but with some edge case failures so it's left undocumented and not officially supported.  Pass a list of devices, e.g. `cuda_device=[0, 1, 2, 3]` instead of a single integer to `Trainer`.  We only implement data parallelism now, and when training with multiple GPUs we split each batch across the GPUs then aggregate gradients before updating parameters.  As a result, when training with multiple GPUs set the batch_size in your configuration should be `num_gpus * batch_size_single_gpu`.

**Comment by larry0123du** [comment]
@matt-peters Cool. For the time being I only need to use the pretrained models so that's what I'm looking for. Any insights on those?

**Comment by matt-peters** [comment]
I've only used this for training so not sure whether it will work out-of-the-box for the pre-trained models.
